pgtoolkit/ctl.py

```diff
@@ -0,0 +1,93 @@
+"""
+.. currentmodule:: pgtoolkit.ctl
+
+API Reference
+-------------
+
+.. autoclass:: PGCtl
+    :members: pg_ctl, init
+.. autofunction:: run_command
+.. autoclass:: CommandRunner
+    :members: __call__
+"""
+
+import functools
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Any, Optional, Sequence, Union, TYPE_CHECKING
+from typing_extensions import Protocol
+
+if TYPE_CHECKING:
+    CompletedProcess = subprocess.CompletedProcess[str]
+else:
+    CompletedProcess = subprocess.CompletedProcess
+
+
+class CommandRunner(Protocol):
+    """Protocol for `run_command` callable parameter of :class:`PGCtl`.
+
+    The `text` mode, as defined in :mod:`subprocess`, must be used in
+    implementations.
+
+    Keyword arguments are expected to match that of :func:`subprocess.run`.
+    """
+
+    def __call__(self, args: Sequence[str], **kwargs: Any) -> CompletedProcess:
+        ...
+
+
+def run_command(args: Sequence[str], **kwargs: Any) -> CompletedProcess:
+    """Default :class:`CommandRunner` implementation for :class:`PGCtl` using
+    :func:`subprocess.run`.
+    """
+    return subprocess.run(
+        args,
+        check=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        universal_newlines=True,
+        **kwargs,
+    )
+
+
+class PGCtl:
+    """Handler for pg_ctl commands.
+
+    :param bindir: location of postgresql user executables; if not specified,
+        this will be determined by calling ``pg_config`` if that executable is
+        found in ``$PATH``.
+    :param run_command: callable implementing :class:`CommandRunner` that will
+        be used to execute ``pg_ctl`` commands.
+    """
+
+    def __init__(
+        self,
+        bindir: Optional[Union[str, Path]] = None,
+        *,
+        run_command: CommandRunner = run_command,
+    ) -> None:
+        if bindir is None:
+            pg_config = shutil.which("pg_config")
```

Comment:
I just tried on a clean Debian buster install and I had to install `postgresql-server-dev-XX` to have `pg_config` correctly running.

If the `dev` package is not installed I get the following message:
```
$ pg_config
You need to install postgresql-server-dev-NN for building a server-side extension or libpq-dev for building a client-side application.
```

Reply:
Then calling `pg_config` in a subprocess later on would crash, I guess (assuming the exit status of `pg_config` is non-zero in that case). I can wrap that error into another `EnvironmentError` to be more explicit.

Reply:
On the same topic, it looks like `pg_config` returns only one `bindir` even if there are several versions of postgresql installed.
```
$ pg_config --bindir
/usr/lib/postgresql/11/bin
$ find /usr/lib/postgresql -name initdb
/usr/lib/postgresql/11/bin/initdb
/usr/lib/postgresql/10/bin/initdb
```

Reply:
The scope of pg_config is limited of its own version. It does not "see" the other versions of PostgreSQL

```
$ /usr/lib/postgresql/12/bin/pg_config --bindir                                                                                                                                
/usr/lib/postgresql/12/bin                                                                                                                                                                                                
$ /usr/lib/postgresql/11/bin/pg_config --bindir                                                                                                                                
/usr/lib/postgresql/11/bin
```

Reply:
That's on purpose. If one does not pass a `bindir` argument, then the default postgres user binaries will be used according to system setup. People wanting to use a specific postgres version should pass a `bindir` argument (or set a specific `pg_config` in their `$PATH`).